### PR TITLE
Bluetooth: Host: Extract out ad_stream to set long adv data

### PR DESCRIPTION
**Pull request by rugeGerritsen**
This structure can be reused to set periodic advertising data.
The structure tries fills the buffer as much as possible.
Later this can be reused for setting advertising data in other cases
as well.

### CODE REVIEW

**Comment by rugeGerritsen** [comment]
FYI. I wanted to extend the advertising extension babblesim test for these changes, but currently the Zephyr LL does not support setting long data. See https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/ll_sw/ull_adv_aux.c#L112

File: subsys/bluetooth/host/adv.c [hunk 1]
```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
...
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	uint8_t current_data_offset;
+};
+
+void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	memset(stream, 0, sizeof(*stream));
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:49]:
```suggestion
	(void)memset(stream, 0, sizeof(*stream));
```

File: subsys/bluetooth/host/adv.c [hunk 2]
```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
...
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:57]: This function is unclear to me. 

I would suggest adding some documentation about the relationship between the `ad_index` and the `ad_len`. 

Alternatively, given how `ad_stream_new` is implemented, couldn't this function return `stream->ad == NULL` instead?

File: subsys/bluetooth/host/adv.c [hunk 3]
```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
...
+
+/* Read data from the flattened AD stream. */
+uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t *len)
+{
+	const struct bt_data *current_data = &stream->ad[stream->ad_index].data[stream->data_index];
+
+	const bool is_first_data = (stream->ad_index == 0) && (stream->data_index == 0);
+	const bool is_last_data = (stream->ad_index == stream->ad_len - 1) &&
+				  (stream->data_index == stream->ad[stream->ad_index].len - 1);
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:66]: Given the multiple uses of `stream->ad_index` and `stream->data_index` I would suggest to extract these as `const` variables and use those instead of deferencing `stream` so many times for the same values.

File: subsys/bluetooth/host/adv.c [hunk 4]
```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
...
+
+	if (stream->current_data_offset == 0) {
+		/* Start reading out a new bt_data structure */
+
+		size_t max_data_to_be_read = current_data->data_len + 2;
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:71]:
```suggestion
		const size_t max_data_to_be_read = current_data->data_len + 2;
```

File: subsys/bluetooth/host/adv.c [hunk 5]
```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
...
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[2], current_data->data, max_data_to_be_read - 2);
+		stream->current_data_offset = max_data_to_be_read - 2;
+
+	} else {
+
+		/* Read the remaining data */
+		ssize_t max_data_to_be_read = current_data->data_len - stream->current_data_offset;
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:89]:
```suggestion
		const ssize_t max_data_to_be_read = current_data->data_len - stream->current_data_offset;
```

File: subsys/bluetooth/host/adv.c [hunk 6]
```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
...
+
+		if (max_data_to_be_read <= *len) {
+			/* The entire bt_data structure fits within the buffer */
+			*len = max_data_to_be_read;
+		} else {
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[2], current_data->data, max_data_to_be_read - 2);
+		stream->current_data_offset = max_data_to_be_read - 2;
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:84]: Given the multiple uses of `- 2` or `+ 2` (which of course is size of the length and type field`, not only in this PR, but in general in the host, we should really consider creating a macro for that value.

File: subsys/bluetooth/host/adv.c [hunk 7]
```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
...
+			*len = max_data_to_be_read;
+		} else {
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[0], current_data->data, max_data_to_be_read);
+		stream->current_data_offset += max_data_to_be_read;
+	}
+
+	uint8_t operation;
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:102]: It is fine to keep as is, but generally we declare variables are top of scope.

File: subsys/bluetooth/host/adv.c [hunk 8]
```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
...
+
+	if (is_first_data && is_last_data &&
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:104]: I suppose that if `is_first_data && is_last_data` then it means that all data fit, right? 

`is_first_data && is_last_data` does read a bit weird, so we could consider storing that in a `const` variable as well (e.g. `is_complete_data`).

File: subsys/bluetooth/host/adv.c [hunk 9]
```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
...
+	} else if (is_first_data) {
+		operation = BT_HCI_LE_EXT_ADV_OP_FIRST_FRAG;
+	} else if (is_last_data && stream->current_data_offset == current_data->data_len) {
+		operation = BT_HCI_LE_EXT_ADV_OP_LAST_FRAG;
+	} else {
+		operation = BT_HCI_LE_EXT_ADV_OP_INTERM_FRAG;
+	}
+
+	/* Increment counters */
+	if (stream->current_data_offset == current_data->data_len) {
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:116]: If this is true, does that mean it is the last fragment? 

Or is there a case where it isn't the last fragment, but this is still true?

**Review by Thalley — CHANGES_REQUESTED** [review]
This looks like a welcome change. 

Some of the code is bit hard to read and understand, so possibly there are some improvements besides my specific comments, but otherwise the code structure looks good.

File: subsys/bluetooth/host/adv.c (hunk 8 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:104]: We can't know if the data is complete before we have appended the data to the buffer. The data may be both first and last, but it is fragmented into multiple parts. Therefore we can't make it const before this evaluation. 

I can add a variable `is_complete_data` right before this if statement, but I'm not sure it will make the code more readable.

File: subsys/bluetooth/host/adv.c (hunk 9 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:116]: It means it was the last fragment of the current advertising data structure. That is not the same as BT_HCI_LE_EXT_ADV_OP_LAST_FRAG as there may be more advertising data structures coming.

Comment (Thalley) [reply subsys/bluetooth/host/adv.c:117]: Right. This is due to the fact that we send the set data command for each LTV structure, right? (Which arguable is still a horribly inefficient way of doing that, but that's outside the scope of this PR). 

But correct me if I am wrong then: 
`is_first_data` seems to be true if it is very first data, and not just the first part of one of the LTV structures. `is_last_data` is however not the last data, but rather the last LTV structure. 

`is_last_data` may then be a bit ill-named if it indeed not the last data as you mention in https://github.com/zephyrproject-rtos/zephyr/pull/41537/files#r778169407

Or did I misunderstand you?

Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:117]: ìs_first_data` and `is_last_data` refer to the LTV structures. Maybe these should be rephrased then? I tried to use the existing terminology.

Comment (Thalley) [reply subsys/bluetooth/host/adv.c:117]: if `is_first_data` is the first LTV, and `is_last_data` also refer to the last LTV, I'd just replace `data` with `ltv` or `ad`.

File: subsys/bluetooth/host/adv.c (hunk 8 above)
Comment (Thalley) [reply subsys/bluetooth/host/adv.c:105]: Ignore this comment; it was based on an incorrect understanding of what `data` meant here.

File: subsys/bluetooth/host/adv.c [hunk 10]
```diff
@@ -27,6 +27,112 @@ enum adv_name_type {
...
+
+		(void)memcpy(&buf[0], current_data->data, max_data_to_be_read);
+		stream->current_data_offset += max_data_to_be_read;
+	}
+
+	if (is_first_ltv && is_last_ltv &&
+	    stream->current_data_offset == current_data->data_len) {
+		operation = BT_HCI_LE_EXT_ADV_OP_COMPLETE_DATA;
+	} else if (is_first_ltv &&
+		   *len == stream->current_data_offset) {
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:109]:
```suggestion
	} else if (is_first_ltv && *len == stream->current_data_offset) {
```

File: subsys/bluetooth/host/adv.c [hunk 11]
```diff
@@ -27,6 +27,112 @@ enum adv_name_type {
...
+		} else {
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[0], current_data->data, max_data_to_be_read);
+		stream->current_data_offset += max_data_to_be_read;
+	}
+
+	if (is_first_ltv && is_last_ltv &&
+	    stream->current_data_offset == current_data->data_len) {
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:106]: The comparison`stream->current_data_offset == current_data->data_len` is done multiple times. 

It is still a bit unclear to me what this exactly means, so perhaps storing the result of this in a variable could be beneficial.

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/adv.c [hunk 12]
```diff
@@ -27,6 +27,112 @@ enum adv_name_type {
...
+	size_t ad_len;
+
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	uint8_t current_data_offset;
+};
+
+void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:47]:
```suggestion
static void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
```

File: subsys/bluetooth/host/adv.c [hunk 13]
```diff
@@ -27,6 +27,112 @@ enum adv_name_type {
...
+{
+	(void)memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+bool ad_stream_is_empty(const struct ad_stream *stream)
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:54]:
```suggestion
static bool ad_stream_is_empty(const struct ad_stream *stream)
```

File: subsys/bluetooth/host/adv.c [hunk 14]
```diff
@@ -27,6 +27,112 @@ enum adv_name_type {
...
+{
+	return stream->ad == NULL;
+}
+
+/* Read data from the flattened AD stream. */
+uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t *len)
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:60]:
```suggestion
static uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t *len)
```

File: subsys/bluetooth/host/adv.c (hunk 11 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c]: Renamed `current_data` -> `current_ltv` to make it more clear

**Comment by rugeGerritsen** [comment]
@Thalley , I've refactored the entire implementation on request of @alwa-nordic . Please re-review.

**Review by alwa-nordic — APPROVED** [review]

File: subsys/bluetooth/host/adv.c [hunk 15]
```diff
@@ -27,6 +27,104 @@ enum adv_name_type {
...
+}
+
+static bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+static const struct bt_data *ad_stream_current_ltv_get(struct ad_stream *stream)
+{
+	const struct bt_data *current_ltv = &stream->ad[stream->ad_index].data[stream->data_index];
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:64]: This is confusing me. Why isn't it 
```suggestion
	const struct bt_data *current_ltv = stream->ad[stream->ad_index].data;
```
instead? `data_index` should not be relevant when retrieving an LTV structure. 

Furthermore, the naming should probably also become persistent. Either use `ad` or `ltv`. Using `ad_index` to get an `ltv` is odd, compared to using `ad_index` to get an `ad` or an `ltv_index` to get an `ltv`.

File: subsys/bluetooth/host/adv.c [hunk 16]
```diff
@@ -27,6 +27,104 @@ enum adv_name_type {
...
+	const bool current_ltv_complete = (stream->current_ltv_offset == current_ltv->data_len + 2);
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:65]: I'm not sure I follow this either. How does comparing an offset to a data length determine if something is complete?

File: subsys/bluetooth/host/adv.c [hunk 17]
```diff
@@ -27,6 +27,104 @@ enum adv_name_type {
...
+			stream->ad_index++;
+		} else {
+			stream->data_index++;
+		}
+	}
+
+	if (ad_stream_is_empty(stream)) {
+		return NULL;
+	} else {
+		return &stream->ad[stream->ad_index].data[stream->data_index];
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:81]: I feel like `stream->ad_index` is used enough to warrant a `const` variable

File: subsys/bluetooth/host/adv.c [hunk 18]
```diff
@@ -27,6 +27,104 @@ enum adv_name_type {
...
+{
+	const struct bt_data *current_ltv = &stream->ad[stream->ad_index].data[stream->data_index];
+	const bool current_ltv_complete = (stream->current_ltv_offset == current_ltv->data_len + 2);
+
+	if (current_ltv_complete) {
+		stream->current_ltv_offset = 0;
+
+		if (stream->data_index + 1 == stream->ad[stream->ad_index].len) {
+			stream->data_index = 0;
+			stream->ad_index++;
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:72]: This function is called `current_ltv_get`, but we increase the `ad_index` here, and return `stream->ad[stream->ad_index]` 
later. 

That means that it does not return the current, but may return either the current or the next, while also modifying the input, which should (almost) never be done for a `get` function.

File: subsys/bluetooth/host/adv.c [hunk 19]
```diff
@@ -27,6 +27,104 @@ enum adv_name_type {
...
+		if (stream->current_ltv_offset == 0) {
+			buf[written_len] = current_ltv->data_len + 1;
+			stream->current_ltv_offset++;
+			written_len++;
+		} else if (stream->current_ltv_offset == 1) {
+			buf[written_len] = current_ltv->type;
+			stream->current_ltv_offset++;
+			written_len++;
+		} else {
+			size_t size_to_copy = MIN(buf_len, current_ltv->data_len);
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:115]:
```suggestion
			const size_t size_to_copy = MIN(buf_len, current_ltv->data_len);
```

File: subsys/bluetooth/host/adv.c [hunk 20]
```diff
@@ -27,6 +27,104 @@ enum adv_name_type {
...
+ *
+ * @param stream  AD stream, @ref ad_stream_new
+ * @param buf     Buffer where the data will be put
+ * @param buf_len Buffer length
+ *
+ * @returns The written length
+ */
+static uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t buf_len)
+{
+	uint8_t written_len = 0;
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:97]: Read instead of write?

```suggestion
 * @returns The read length
 */
static uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t buf_len)
{
	uint8_t read_len = 0;
```

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/adv.c (hunk 15 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:64]: It is because `stream->ad` is a two-dimensional array. `stream->ad` contains `stream->ad_len` elements of type `struct bt_ad`. Each `struct bt_ad` consists of `stream->ad[i].len` LTV groups (`struct bt_data`).

File: subsys/bluetooth/host/adv.c (hunk 16 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:65]: If we have read out the entire LTV, it is complete. The data_len is the data length of the LTV. The offset is the amount of data read including type and length field (as documented)

File: subsys/bluetooth/host/adv.c (hunk 17 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:81]: Do you mean the following?
```
const size_t ad_index = stream->ad_index;
const size_t data_index = stream->data_index;

return &stream->ad[ad_index].data[data_index];
```
I don't think this improves the readability.

File: subsys/bluetooth/host/adv.c (hunk 18 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:72]: What about naming it `ad_stream_current_ltv_update()`, or do you have a proposal for what and how to change it? I can simply copy out the content of this extracted function and paste it in the `_read`. I don't think that makes the code more readable.

File: subsys/bluetooth/host/adv.c (hunk 20 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:97]: I guess that is a matter of semantics. I can change it if you want

File: subsys/bluetooth/host/adv.c (hunk 17 above)
Comment (Thalley) [reply subsys/bluetooth/host/adv.c:81]: Yes. 

It may not necessarily increase readability, but depending on the compiler, it may be more efficient (as you will have fewer pointere deferences). 

It is OK to keep, but I'd personally favor storing the value in a variable.

File: subsys/bluetooth/host/adv.c (hunk 15 above)
Comment (Thalley) [reply subsys/bluetooth/host/adv.c:64]: Alright, I had misunderstood the dimensions then. 

I figured it was an array of `struct bt_ad` which in turn is a header and an array of data, but you are saying that we have an array of `struct bt_ad` which each has an array of `struct bt_data` which each has an array of data (alongside the header), effectively being a 3 dimensional array if you include the `data` array in `struct bt_data`. 

Can you elaborate on why it is two-dimensional array then? The API takes a single array of `struct bt_data`, so why are we adding an additional dimension? 

Another issue, which is course not introduce by this PR, is that we use `data` in `bt_ad` but also in `bt_data`, so it becomes unclear which `data` that `data_index` refers to, which is also causing some of the confusion I suppose.

File: subsys/bluetooth/host/adv.c (hunk 16 above)
Comment (Thalley) [reply subsys/bluetooth/host/adv.c:65]: I'm sorry, but I don't understand that. 

```

	/* Current LTV offset contains the offset in the AD structure.
	 * The length and type are included in this offset.
	 */
```

`Current LTV offset contains the offset in the AD structure.` - What is the offset in this case? As is, what is it offsetting? 

Let's assume the simplest case, of a single LTV structure being set in the advertising data, where the value length (data_len) is 5.  
In that case, I'd expect `current_ltv_offset` to start at 0, but where clearly it is a complete LTV structure, but the comparison here would return false, as `0 != 5 + 2`. 

Can you please help me understand this better?

File: subsys/bluetooth/host/adv.c (hunk 18 above)
Comment (Thalley) [reply subsys/bluetooth/host/adv.c:72]: `ad_stream_current_ltv_update` sounds like a decent solution; as least the caller would expect the input to change then.

File: subsys/bluetooth/host/adv.c (hunk 15 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:64]: The `struct bt_ad` is used to add the complete local name to the advertising data without doing additional copying. See https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/adv.c#L650. 

In my opinion it would be more readable if we change then name of `struct bt_ad` to `struct bt_ad_list` or something similar.

File: subsys/bluetooth/host/adv.c (hunk 16 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:65]: For the case of your example, the advertising data is `[L, T, <5 bytes of data>]`. The `current_ltv_offset` indicates how much of this data that has been read. It starts out at 0, is incremented to 1 when the length field is read, and ends up at 7 when length, type, and value have been read.

`current_ltv_complete` here refers to if the reading of it is complete. Maybe it should be renamed to `done_reading_ltv` or something similar?

File: subsys/bluetooth/host/adv.c (hunk 15 above)
Comment (Thalley) [reply subsys/bluetooth/host/adv.c:64]: Perhaps `struct bt_data_list` (or `array` instead of `list`), as that is what is really is, right?

Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:64]: Yes, something similar to that. In my opinion `data` is sometimes a bit too generic. But that is my personal opinion

File: subsys/bluetooth/host/adv.c (hunk 16 above)
Comment (Thalley) [reply subsys/bluetooth/host/adv.c:65]: Ahhh, Eureka moment :D 

So what we are doing here, is that we are incrementing the `ad_index` or the `data_index` in the beginning of the function if the previous call of the function finished an LTV. 

Since `current_ltv_offset` is updated outside the scope of this function, we can't do that increment at the end of this function. **But does also mean that in order to parse/read a single LTV structure, this function would need to be called twice?** 

Not sure what feedback @alwa-nordic gave regarding the refactoring of this, but since he approved it, this must have been what he wanted, although it does feel somewhat odd that we have a function `ad_stream_current_ltv_get` that depend on a caller `ad_stream_read` incrementing a specific struct field between calls. 

At this point, I think it was easier to read and understand before the refactoring, but with enough added comments and documentation for how these two (highly coupled) function need to work together, this would be OK.

Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:65]: > But does also mean that in order to parse/read a single LTV structure, this function would need to be called twice

Yes you are right. 

The benefit is now that it can read more than one LTV at the time. We should there be able to reuse this function in other places as well given that we incorporate the shortening of advertising data name.

File: subsys/bluetooth/host/adv.c (hunk 15 above)
Comment (Thalley) [reply subsys/bluetooth/host/adv.c:64]: It is very generic :) I assume that `ad` also refers to advertising data, which is also very generic, and `bt_data` actually referring to LTV-structured data, which is slightly less generic. 

I think it is OK to keep the naming as is for now, as changing it would be significantly outside the scope of this PR, but we could add comments for the structs and fields to improve the understanding.

Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:64]: Feel free to add suggestions for better comments to the structure members

File: subsys/bluetooth/host/adv.c [hunk 21]
```diff
@@ -27,6 +27,121 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:33]:
```suggestion
struct bt_ad {
	/* Pointer to an LTV structure */
	const struct bt_data *data;
	/* Number of elements in @p data */
	size_t len;
};
```

File: subsys/bluetooth/host/adv.c [hunk 22]
```diff
@@ -27,6 +27,121 @@ enum adv_name_type {
...
+	size_t ad_len;
+
+	/* The current index in the array of struct bt_ad elements */
+	size_t ad_index;
+	/* The current index in the array of ad.data elements */
+	size_t data_index;
+
+	/* Current LTV offset contains the offset in the AD structure.
+	 * The length and type are included in this offset.
+	 */
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:48]:
```suggestion
	/* Current LTV offset contains the data offset in the ad[x].data[y].data value array
	 * The length and type are included in this offset.
	 */
```

File: subsys/bluetooth/host/adv.c [hunk 23]
```diff
@@ -27,6 +27,121 @@ enum adv_name_type {
...
+	(void)memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+/** @brief Returns true if the current stream is empty.
+ *
+ * @param stream AD stream, @ref ad_stream_new
+ *
+ *
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:63]:
```suggestion
 *
```

File: subsys/bluetooth/host/adv.c [hunk 24]
```diff
@@ -27,6 +27,121 @@ enum adv_name_type {
...
+ * @returns true if the stream is now empty.
+ */
+static bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:69]: Why was this changed back to index/len comparison, instead of a `NULL` check?

File: subsys/bluetooth/host/adv.c [hunk 25]
```diff
@@ -27,6 +27,121 @@ enum adv_name_type {
...
+
+static void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	(void)memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+/** @brief Returns true if the current stream is empty.
+ *
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:60]:
```suggestion
/** 
 * @brief Returns true if the current stream is empty.
 *
```

File: subsys/bluetooth/host/adv.c [hunk 26]
```diff
@@ -27,6 +27,121 @@ enum adv_name_type {
...
+ * @returns true if the stream is now empty.
+ */
+static bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+/** @brief Returns the bt_data structure that is currently being read
+ *
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:72]:
```suggestion
/** 
 * @brief Returns the bt_data structure that is currently being read
 *
```

File: subsys/bluetooth/host/adv.c [hunk 27]
```diff
@@ -27,6 +27,121 @@ enum adv_name_type {
...
+ * The read data can contain multiple LTV AD structures.
+ *
+ * @param stream  AD stream, @ref ad_stream_new
+ * @param buf     Buffer where the data will be put
+ * @param buf_len Buffer length
+ *
+ * @returns The number of bytes read from the stream written to the provided buffer
+ */
+static uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t buf_len)
```
Comment (Thalley) [inline subsys/bluetooth/host/adv.c:112]: Suggestion: We could actually also have used the `struct net_buf_simple` instead of supplying a `buf`, `buf_len` and returning the number of bytes read, as the `net_buf_simple` API could do that for us. 

It is OK to keep as is, but something to consider.

File: subsys/bluetooth/host/adv.c (hunk 15 above)
Comment (Thalley) [reply subsys/bluetooth/host/adv.c:81]: Added a few suggestions

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/adv.c (hunk 27 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:112]: I will keep it as is for now. The PR for allowing to set long periodic adv data has been open for two months now, I would like to get the functionality merged. I see that this structure can be reused in other places. I suggest to do that refactoring then

File: subsys/bluetooth/host/adv.c (hunk 24 above)
Comment (rugeGerritsen) [reply subsys/bluetooth/host/adv.c:69]: When refactoring it caused `ad_stream_current_ltv_update()` to look a bit ugly. Discussed with @alwa-nordic

Comment (Thalley) [reply subsys/bluetooth/host/adv.c:69]: Then please add a comment for this function. 

While it makes sense for me right now, it was confusing to me at first, so will likely be that for me again later, or for others. 

Something along the line of `if ad_index == ad_len, then we are past the last element in the ad array`

**Review by Thalley — APPROVED** [review]

**Comment by rugeGerritsen** [comment]
Thanks for reviewing. The new code is more efficient and reusable than the initial proposal.

**Review by jhedberg — APPROVED** [review]
